pkg/agent/core/state.go

```diff
@@ -791,12 +852,54 @@ func (s *state) desiredResourcesFromMetricsOrRequestedUpscaling(now time.Time) (
 			return nil
 		}
 	}
+	s.updateTargetRevision(now, result, s.VM.Using())
+
+	// TODO: we are both saving the result into LastDesiredResources and returning it. This is
+	// redundant, and we should remove one of the two.
+	s.LastDesiredResources = &result
 
-	s.info("Calculated desired resources", zap.Object("current", s.VM.Using()), zap.Object("target", result))
+	s.info("Calculated desired resources",
+		zap.Object("current", s.VM.Using()),
+		zap.Object("target", result),
+		zap.Object("targetRevision", &s.TargetRevision))
 
 	return result, calculateWaitTime
 }
 
+func (s *state) updateTargetRevision(now time.Time, desired api.Resources, current api.Resources) {
+	if s.LastDesiredResources == nil {
+		s.LastDesiredResources = &current
+	}
+
+	if *s.LastDesiredResources == desired {
+		// Nothing changed, so no need to update the target revision
+		return
+	}
+
+	var flags vmv1.Flag
+
+	if desired.HasFieldGreaterThan(*s.LastDesiredResources) {
+		flags.Set(revsource.Upscale)
+	}
+	if desired.HasFieldLessThan(*s.LastDesiredResources) {
+		flags.Set(revsource.Downscale)
+	}
+
+	s.TargetRevision = s.Config.RevisionSource.Next(now, flags)
+}
+
+func (s *state) updateNeonVMCurrentRevision(currentRevision vmv1.RevisionWithTime) {
+	revsource.Propagate(currentRevision.UpdatedAt.Time,
+		s.NeonVM.TargetRevision,
+		&s.NeonVM.CurrentRevision,
+		s.Config.ObservabilityCallbacks.NeonVMLatency,
+	)
+	err := s.Config.RevisionSource.Observe(currentRevision.UpdatedAt.Time, currentRevision.Revision)
+	if err != nil {
+		s.warnf("Failed to observe clock source: %v", err)
+	}
```

Comment:
If we get an error here, do you think it's likely that the error is due to some persistent bad state?

If so, can we change this to panic? that will reset all of the internal state that the agent tracks for this VM

Reply:
> If we get an error here, do you think it's likely that the error is due to some persistent bad state?

Not necessarily. For example, the agent might've restarted, and we are getting the revisions generated by a previous agent.

Even if I initialize the revision source with last known revision, it can still race.

Also, revision in status is "external input", meaning something external could've change that, and I think we shouldn't panic on external input.

Reply:
> Not necessarily. For example, the agent might've restarted, and we are getting the revisions generated by a previous agent.
>
> Even if I initialize the revision source with last known revision, it can still race.

Sure, yeah -- I'm just thinking that if we get in a bad state that we can't recover from, it would be better to reset the state. But if you think that's unlikely relative to the amount of false-positives we'd get around agent restarts, let's leave as is.

> I think we shouldn't panic on external input

For the sake of general discussion: In most other contexts I would agree with you; here however, panicking triggers a clean restart of _just_ the tasks associated with a particular VM. IMO it's a clean way to recover from a degraded state.

Reply:
Anyway, the recovery here should happen once autoscaler-agent writes fresh `targetRevision`, and it is propagated.